common/src/main/java/io/crate/types/RowType.java

```diff
@@ -0,0 +1,162 @@
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.types;
+
+import io.crate.Streamer;
+import io.crate.data.Row;
+import io.crate.data.RowN;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class RowType extends DataType<Row> implements Streamer<Row> {
+
+    public static final int ID = 18;
+    public static final String NAME = "record";
+    private final List<DataType<?>> fieldTypes;
+    private final List<String> fieldNames;
+
+    /**
+     * @param fieldTypes the types of the fields within a Row
+     * @param fieldNames the names of the fields. This can be empty in which case defaults will be created (colX ... colN)
+     */
+    public RowType(List<DataType<?>> fieldTypes, List<String> fieldNames) {
+        assert fieldNames.isEmpty() || fieldNames.size() == fieldTypes.size()
+            : "fieldNames must either be empty or have the same length as fieldTypes";
+        this.fieldTypes = fieldTypes;
+        if (fieldNames.isEmpty() && !fieldTypes.isEmpty()) {
+            ArrayList<String> generatedFieldNames = new ArrayList<>(fieldTypes.size());
+            for (int i = 0; i < fieldTypes.size(); i++) {
+                generatedFieldNames.add("col" + (i + 1));
+            }
+            this.fieldNames = List.copyOf(generatedFieldNames);
+        } else {
+            this.fieldNames = List.copyOf(fieldNames);
+        }
+    }
+
+    public RowType(List<DataType<?>> fieldTypes) {
+        this(fieldTypes, List.of());
+    }
+
+    public RowType(StreamInput in) throws IOException {
+        int numFields = in.readVInt();
+        ArrayList<DataType<?>> fieldTypes = new ArrayList<>(numFields);
+        ArrayList<String> fieldNames = new ArrayList<>(numFields);
+        for (int i = 0; i < numFields; i++) {
+            fieldTypes.add(DataTypes.fromStream(in));
+            fieldNames.add(in.readString());
+        }
+        this.fieldTypes = List.copyOf(fieldTypes);
+        this.fieldNames = List.copyOf(fieldNames);
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(fieldTypes.size());
+        for (int i = 0; i < fieldTypes.size(); i++) {
+            DataTypes.toStream(fieldTypes.get(i), out);
+            out.writeString(fieldNames.get(i));
+        }
+    }
+
+    public int numElements() {
+        return fieldTypes.size();
+    }
+
+    public DataType<?> getFieldType(int position) {
+        return fieldTypes.get(position);
+    }
+
+    public String getFieldName(int position) {
```

Comment:
should we have some pre condition here to be more safe and get a better exception message if we get out of bound exception.

Reply:
Not sure, I think the out-of-bounds exception that will be raised is fairly fitting.